### DISCUSSION TITLE
Virtual branch like GitButler

### DISCUSSION THREAD

**Discussion by black7375** [arxanas/git-branchless#1257] [Ideas]
- https://gitbutler.com/

I'm impressed that you can create virtual branches, staging and commit, submit to each virtual branch.
I'm curious to know what you think.

**Comment by JaniM** [comment]
Virtual branches are very clever and feel very nice to use, but they are difficult to integrate with other git tooling. Git-branchless is nice in that it is entirely invisible to git.

As for the workflow, I find that stacked commits/branches solve most of what I'd want to use GitButler for. If I am working on A, I can have a commit for it. If I fix a bug while working on it, I can fix it and place the fix commit before A. Then before releasing my changes, if the changes are independent, I can move A to main again and push it. If not, I make stacked PRs.

**Reply by black7375** [reply]
Certainly `git prev/next` and `git amend` solve a lot of problems.

**Reply by hauleth** [reply]
My main goal is that I want to be able to push stuff without moving it to `origin/HEAD`. Reason for that is that sometimes I have independent work *on top* of the previous commit. As an example - I sometimes need to fix test that is failing because of my unusual setup. After that I go back to work on whatever I was working previously on top of that change. However I want to continue my work and create independent PR for that set of changes. It would be *way* easier to implement such workflow if there would be `git branchless copy` or `git branchless move --no-hide/--copy` that would do in-memory cherry-picking, as then I would be able to move some commits to that "hidden PR branch" that would be then sent to GitHub as a PR base.

**Reply by arxanas** [reply]
The virtual branch workflow would have been useful for me for some situations, but perhaps primarily because managing an explicit merge commit would be onerous. jj has several features that support a "virtual branch" workflow by creating an automatically-updated merge commit. Neither git-branchless nor jj support the "absorb" feature to automatically amend changes into one of the ancestor commits, which is one example of a useful workflow not currently supported by tooling.

Nonetheless, other tooling doesn't support the merge commits well, which makes it hard to use them in practice even when local tooling supports it. For example Phabricator can render DAG code review structures in the UI, but offers no way to automatically create the DAG structure from the local commit topology, nor a way to land those commits automatically in topological order. (Perhaps Gerrit fares better?) 

In practice, I have just declared linear dependency structures and reordered non-dependent commits so that I could land the earliest-accepted ones first.

**Reply by black7375** [reply]
Definitely `git amend` and `git restack` are killer features for me. thank you
